README.md
# ROMAN_NUMERAL_PROJECT
Roman Numeral Calculator Program - HSSAMRA


change_log.txt
***** CHANGELOG *****

-> 02/06/2017 (branch: feature_add_version_incrementation)
	- added main.c, common.h, inc_bld_str.sh, Makefile, and bld_num.txt to project
	- added feature to increment build number everytime code is compiled
	- added feature to main.c to display software version number when started
	- deleted "feature_add_version_incrementation" branch after merging it to "develop" branch
	
-> 02/06/2017 (branch: feature_verify_input_string)
	- added change_log.txt, rmn_calc.c, rmn_calc.h, test_rmn_calc.c, and test_rmn_calc.h to project
	- edited Makefile to include necessary files (rmn_calc.c, test_rmn_calc.c)
	- verified regex works for case insensitivity
	- verified regex matches for 0 - 3 'i' characters and fails if there are more
	- verified regex matches for 0 - 1 'v' characters and fails if there are more
	- verified regex matches for 'vi' character combination and fails if 'v' comes after 'i' ('viv')
	
-> 02/07/2017 (branch: feature_verify_input_string)
	- removed file path variable from inc_bld_str.sh since it will be different on client machines
	- verified regex matches for 0 - 3 'x' characters and fails if there are more
	- verified regex matches for 0 - 1 'l' characters and fails if there are more
	- verified regex matches for 0 - 3 'c' characters and fails if there are more
	- verified regex matches for 0 - 1 'd' characters and fails if there are more
	- verified regex matches for 0 - n 'm' characters
	- verified regex matches edge conditions for 'v' 'i' combinations and 'x' 'i' combinations
	- verified regex matches edge conditions for 'iv' using edge conditions
	- verified regex matches edge conditions for 'ix'

-> 02/08/2017 (branch: feature_verify_input_string)
	- verified regex matches edge conditions for 'xl'
	- verified regex matches edge conditions for 'xc'
	- verified regex matches edge conditions for 'cd'
	- verified regex matches edge conditions for 'cm'
	- merging this branch back into develop in order to work on the next feature:
		- converting integers into roman numerals
		- this will allow for full testing of string verification in the long run

-> 02/08/2017 (branch: feature_encode_int_to_roman_numeral)
	- created a new branch in order to take an integer and create a roman numeral string
	- renamed rmn_calc.c to test_rmn_calc.c to differentiate between different tests
	- added test_rmn_calc_enc_dcd.c to project for encoding and decoding roman numeral strings
	- verified encoding to 'I' for edge conditions 1 and 3
	- verified encoding to 'IV' for 4
	- verified encoding to 'V' for edge conditions 5 and 8
	- verified encoding to 'IX' for 9
	- verified encoding to 'X' for edge conditions 10 and 39
	- verified encoding to 'XL' for edge conditions 40 and 49
	- verified encoding to 'L' for edge conditions 50 and 89
	- verified encoding to 'XC' for edge conditions 90 and 99
	- verified encoding to 'C' for edge conditions 100 and 399
	- verified encoding to 'CD' for edge conditions 400 and 499
	- verified encoding to 'D' for edge conditions 500 and 899
	- verified encoding to 'CM' for edge conditions 900 and 999
	- verified encoding to 'M' for conditions of 1000 and 1999
	
-> 02/08/2017 (branch: feature_verify_input_string)
	- in process of adding test to check file that the roman numerals are written to and compare it to a master file to verify accuracy

-> 02/09/2017 (branch: feature_verify_input_string)
	- added test for writing generated roman numerals to file after verification, and compare to master file to verify accuracy
	
-> 02/09/2017 (branch: feature_decode_roman_numeral_to_int)
	- added new branch to add feature needed to decode roman numerals to integers
	- verified decoding to 1 for 'I'
	

test_rmn_calc.c
// --------------- test_rmn_calc.c Log --------------------
//
// 02/06/2017	HSamra	Initial Commit
// 02/06/2017	HSamra	Adding tests for regex comparisons
//							- added tests for i and I (min: 0, max: 3)
//							- added tests for v and V (min: 0, max: 1)
//							- added tests for vi
//
// --------------------------------------------------------

#include <assert.h>
#include "common.h"

void _regex_verifyI_pass(char *str)
{
	printf("Test_regex_verifyI_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyI_fail(char *str)
{
	printf("Test_regex_verifyI_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyV_pass(char *str)
{
	printf("Test_regex_verifyV_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyV_fail(char *str)
{
	printf("Test_regex_verifyV_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyX_pass(char *str)
{
	printf("Test_regex_verifyX_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyX_fail(char *str)
{
	printf("Test_regex_verifyX_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyL_pass(char *str)
{
	printf("Test_regex_verifyL_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyL_fail(char *str)
{
	printf("Test_regex_verifyL_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyC_pass(char *str)
{
	printf("Test_regex_verifyC_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyC_fail(char *str)
{
	printf("Test_regex_verifyC_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyD_pass(char *str)
{
	printf("Test_regex_verifyD_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyD_fail(char *str)
{
	printf("Test_regex_verifyD_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyM_pass(char *str)
{
	printf("Test_regex_verifyM_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_VI_pass(char *str)
{
	printf("Test_regex_verifyMulti_VI_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_VI_fail(char *str)
{
	printf("Test_regex_verifyMulti_VI_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_XI_pass(char *str)
{
	printf("Test_regex_verifyMulti_XI_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_XI_fail(char *str)
{
	printf("Test_regex_verifyMulti_XI_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_XV_pass(char *str)
{
	printf("Test_regex_verifyMulti_XV_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_XV_fail(char *str)
{
	printf("Test_regex_verifyMulti_XV_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_all_pass(char *str)
{
	printf("Test_regex_verifyMulti_all_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_all_fail(char *str)
{
	printf("Test_regex_verifyMulti_all_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_IV_pass(char *str)
{
	printf("Test_regex_verifySubTerm_IV_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_IV_fail(char *str)
{
	printf("Test_regex_verifySubTrm_IV_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_IX_pass(char *str)
{
	printf("Test_regex_verifySubTrm_IX_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_IX_fail(char *str)
{
	printf("Test_regex_verifySubTrm_IX_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_XL_pass(char *str)
{
	printf("Test_regex_verifySubTrm_XL_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_XL_fail(char *str)
{
	printf("Test_regex_verifySubTrm_XL_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_XC_pass(char *str)
{
	printf("Test_regex_verifySubTrm_XC_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_XC_fail(char *str)
{
	printf("Test_regex_verifySubTrm_XC_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_CD_pass(char *str)
{
	printf("Test_regex_verifySubTrm_CD_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_CD_fail(char *str)
{
	printf("Test_regex_verifySubTrm_CD_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}


void _regex_verifySubTrm_CM_pass(char *str)
{
	printf("Test_regex_verifySubTrm_CM_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_CM_fail(char *str)
{
	printf("Test_regex_verifySubTrm_CM_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void TEST_rmn_calc_test(void)
{
	// test conditions for each character
	_regex_verifyI_pass("I");
	_regex_verifyI_pass("i");
	_regex_verifyI_pass("ii");
	_regex_verifyI_pass("iii");
	_regex_verifyI_fail("iiii");
	
	_regex_verifyV_pass("V");
	_regex_verifyV_pass("v");
	_regex_verifyV_fail("vv");
	
	_regex_verifyX_pass("X");
	_regex_verifyX_pass("x");
	_regex_verifyX_pass("xx");
	_regex_verifyX_pass("xxx");
	_regex_verifyX_fail("xxxx");
	
	_regex_verifyL_pass("L");
	_regex_verifyL_pass("l");
	_regex_verifyL_fail("ll");

	_regex_verifyC_pass("C");
	_regex_verifyC_pass("c");
	_regex_verifyC_pass("cc");
	_regex_verifyC_pass("ccc");
	_regex_verifyC_fail("cccc");
	
	_regex_verifyD_pass("D");
	_regex_verifyD_pass("d");
	_regex_verifyD_fail("dd");
	
	_regex_verifyM_pass("M");
	_regex_verifyM_pass("m");
	_regex_verifyM_pass("mmmmmm");
	
	// testing edge conditions
	_regex_verifyMulti_VI_pass("vi");
	_regex_verifyMulti_VI_fail("viiii");
	_regex_verifyMulti_VI_fail("vvi");
	
	_regex_verifyMulti_XI_pass("xi");
	_regex_verifyMulti_XI_pass("xxxiii");
	_regex_verifyMulti_XI_fail("xiiii");
	_regex_verifyMulti_XI_fail("xxxxi");
	
	_regex_verifyMulti_XV_pass("xv");
	_regex_verifyMulti_XV_pass("xxxv");
	_regex_verifyMulti_XV_fail("xvv");
	_regex_verifyMulti_XV_fail("xxxxv");
	
	_regex_verifyMulti_all_pass("mdclxvi");
	_regex_verifyMulti_all_fail("mcdivxl");
	
	// test conditions for subtractive terms (iv, ix, xl, xc, cd, cm)
	_regex_verifySubTrm_IV_pass("iv");
	_regex_verifySubTrm_IV_fail("ivi");
	_regex_verifySubTrm_IV_fail("ivv");
	_regex_verifySubTrm_IV_fail("viv");
	
	_regex_verifySubTrm_IX_pass("ix");
	_regex_verifySubTrm_IX_pass("xix");
	_regex_verifySubTrm_IX_fail("ixi");
	_regex_verifySubTrm_IX_fail("ixiv");
	_regex_verifySubTrm_IX_fail("ixv");
	
	_regex_verifySubTrm_XL_pass("xl");
	_regex_verifySubTrm_XL_pass("xli");
	_regex_verifySubTrm_XL_pass("xlix");
	_regex_verifySubTrm_XL_pass("cxl");
	_regex_verifySubTrm_XL_fail("xlx");
	_regex_verifySubTrm_XL_fail("xll");
	
	_regex_verifySubTrm_XC_pass("xc");
	_regex_verifySubTrm_XC_pass("cxc");
	_regex_verifySubTrm_XC_pass("xcix");
	_regex_verifySubTrm_XC_pass("xcvi");
	_regex_verifySubTrm_XC_pass("xciv");
	_regex_verifySubTrm_XC_fail("xcx");
	_regex_verifySubTrm_XC_fail("xcxl");
	_regex_verifySubTrm_XC_fail("xcl");
	
	_regex_verifySubTrm_CD_pass("cd");
	_regex_verifySubTrm_CD_pass("cdxl");
	_regex_verifySubTrm_CD_pass("cdxc");
	_regex_verifySubTrm_CD_pass("cdi");
	_regex_verifySubTrm_CD_pass("cdv");
	_regex_verifySubTrm_CD_pass("cdx");
	_regex_verifySubTrm_CD_fail("cdc");
	_regex_verifySubTrm_CD_fail("dcd");
	
	_regex_verifySubTrm_CM_pass("cm");
	_regex_verifySubTrm_CM_pass("cmi");
	_regex_verifySubTrm_CM_pass("cmxc");
	_regex_verifySubTrm_CM_pass("cmxl");
	_regex_verifySubTrm_CM_pass("mcm");
	_regex_verifySubTrm_CM_pass("cmlxvi");
	_regex_verifySubTrm_CM_fail("cmc");
	_regex_verifySubTrm_CM_fail("cmd");
	_regex_verifySubTrm_CM_fail("cmcd");
	
}


test_rmn_calc.h
#ifndef TEST_RMN_CALC_H
#define TEST_RMN_CALC_H

void TEST_rmn_calc_test_regex(void);	// test_rmn_calc_regex.c
void TEST_rmn_calc_test_enc_dcd(void);	// test_rmn_calc_enc_dcd.c

#endif // TEST_RMN_CALC_H


test_rmn_calc_regex.c
// --------------- test_rmn_calc_regex.c Log --------------------
//
// 02/06/2017	HSamra	Initial Commit
// 02/06/2017	HSamra	Adding tests for regex comparisons
//							- added tests for i and I (min: 0, max: 3)
//							- added tests for v and V (min: 0, max: 1)
// 02/07/2017	HSamra	Adding tests for regex comparisions
//							- added tests for x and X (min: 0, max: 3)
//							- added tests for l and L (min: 0, max: 1)
//							- added tests for c and C (min: 0, max: 3)
//							- added tests for d and D (min: 0, max: 1)
//							- added tests for m and M (min: 0, max: 3)
//							- added tests for iv
//							- added tests for ix
// 02/08/2017	HSamra	Adding tests for regex comparisons
//							- added tests for xl
//							- added tests for xc
//							- added tests for cd
//							- added tests for cm
//							- renamed file from test_rmn_calc.c to test_rmn_calc_regex.c
//							- in process of adding test to check file that the roman numerals are written to and compare it to a master file to verify accuracy
// 02/08/2017	HSamra	Adding tests for regex comparisons
//							- added test for writing generated roman numerals to file after verification, and compare to master file to verify accuracy
//
// --------------------------------------------------------

#include <assert.h>
#include "common.h"

const char testFile[] = "temp_test.txt";
const char chckFile[] = "roman_numeral_test.txt";

void _regex_verifyI_pass(char *str)
{
	printf("Test_regex_verifyI_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyI_fail(char *str)
{
	printf("Test_regex_verifyI_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyV_pass(char *str)
{
	printf("Test_regex_verifyV_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyV_fail(char *str)
{
	printf("Test_regex_verifyV_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyX_pass(char *str)
{
	printf("Test_regex_verifyX_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyX_fail(char *str)
{
	printf("Test_regex_verifyX_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyL_pass(char *str)
{
	printf("Test_regex_verifyL_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyL_fail(char *str)
{
	printf("Test_regex_verifyL_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyC_pass(char *str)
{
	printf("Test_regex_verifyC_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyC_fail(char *str)
{
	printf("Test_regex_verifyC_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyD_pass(char *str)
{
	printf("Test_regex_verifyD_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyD_fail(char *str)
{
	printf("Test_regex_verifyD_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyM_pass(char *str)
{
	printf("Test_regex_verifyM_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_VI_pass(char *str)
{
	printf("Test_regex_verifyMulti_VI_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_VI_fail(char *str)
{
	printf("Test_regex_verifyMulti_VI_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_XI_pass(char *str)
{
	printf("Test_regex_verifyMulti_XI_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_XI_fail(char *str)
{
	printf("Test_regex_verifyMulti_XI_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_XV_pass(char *str)
{
	printf("Test_regex_verifyMulti_XV_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_XV_fail(char *str)
{
	printf("Test_regex_verifyMulti_XV_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_all_pass(char *str)
{
	printf("Test_regex_verifyMulti_all_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyMulti_all_fail(char *str)
{
	printf("Test_regex_verifyMulti_all_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_IV_pass(char *str)
{
	printf("Test_regex_verifySubTerm_IV_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_IV_fail(char *str)
{
	printf("Test_regex_verifySubTrm_IV_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_IX_pass(char *str)
{
	printf("Test_regex_verifySubTrm_IX_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_IX_fail(char *str)
{
	printf("Test_regex_verifySubTrm_IX_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_XL_pass(char *str)
{
	printf("Test_regex_verifySubTrm_XL_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_XL_fail(char *str)
{
	printf("Test_regex_verifySubTrm_XL_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_XC_pass(char *str)
{
	printf("Test_regex_verifySubTrm_XC_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_XC_fail(char *str)
{
	printf("Test_regex_verifySubTrm_XC_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_CD_pass(char *str)
{
	printf("Test_regex_verifySubTrm_CD_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_CD_fail(char *str)
{
	printf("Test_regex_verifySubTrm_CD_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}


void _regex_verifySubTrm_CM_pass(char *str)
{
	printf("Test_regex_verifySubTrm_CM_pass(%s): ", str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifySubTrm_CM_fail(char *str)
{
	printf("Test_regex_verifySubTrm_CM_fail(%s): ", str);
	
	assert(RMN_verify_string(str) == 0);
	
	printf("PASS\r\n");
}

void _regex_verifyStr_pass(int n, char *str)
{
	printf("Test_regex_verifyStr_pass(%d -> %s): ", n, str);
	
	assert(RMN_verify_string(str) == 1);
	
	printf("PASS\r\n");
}

void _regex_verifyFile_pass(void)
{
	FILE *f1 = fopen(testFile, "r");
	FILE *f2 = fopen(chckFile, "r");
	
	int fCheck1 = getc(f1); // testFile
	int fCheck2 = getc(f2); // chckFile
	
	printf("Test_regex_verifyFile_pass: ");
	
	while((fCheck1 != EOF) && (fCheck2 != EOF) && (fCheck1 == fCheck2))
	{
		fCheck1 = getc(f1);
		fCheck2 = getc(f2);
	}
	
	fclose(f1);
	fclose(f2);
	
	assert(fCheck1 == fCheck2);
	
	printf("PASS\r\n");
}

void TEST_rmn_calc_test_regex(void)
{
	FILE *file1 = NULL;						 
	int i = 0;
	
	// test conditions for each character
	_regex_verifyI_pass("I");
	_regex_verifyI_pass("i");
	_regex_verifyI_pass("ii");
	_regex_verifyI_pass("iii");
	_regex_verifyI_fail("iiii");
	
	_regex_verifyV_pass("V");
	_regex_verifyV_pass("v");
	_regex_verifyV_fail("vv");
	
	_regex_verifyX_pass("X");
	_regex_verifyX_pass("x");
	_regex_verifyX_pass("xx");
	_regex_verifyX_pass("xxx");
	_regex_verifyX_fail("xxxx");
	
	_regex_verifyL_pass("L");
	_regex_verifyL_pass("l");
	_regex_verifyL_fail("ll");

	_regex_verifyC_pass("C");
	_regex_verifyC_pass("c");
	_regex_verifyC_pass("cc");
	_regex_verifyC_pass("ccc");
	_regex_verifyC_fail("fnamecccc");
	
	_regex_verifyD_pass("D");
	_regex_verifyD_pass("d");
	_regex_verifyD_fail("dd");
	
	_regex_verifyM_pass("M");
	_regex_verifyM_pass("m");
	_regex_verifyM_pass("mmmmmm");

	// testing edge conditions
	_regex_verifyMulti_VI_pass("vi");
	_regex_verifyMulti_VI_fail("viiii");
	_regex_verifyMulti_VI_fail("vvi");
	
	_regex_verifyMulti_XI_pass("xi");
	_regex_verifyMulti_XI_pass("xxxiii");
	_regex_verifyMulti_XI_fail("xiiii");
	_regex_verifyMulti_XI_fail("xxxxi");
	
	_regex_verifyMulti_XV_pass("xv");
	_regex_verifyMulti_XV_pass("xxxv");
	_regex_verifyMulti_XV_fail("xvv");
	_regex_verifyMulti_XV_fail("xxxxv");
	
	_regex_verifyMulti_all_pass("mdclxvi");
	_regex_verifyMulti_all_fail("mcdivxl");
	
	// test conditions for subtractive terms (iv, ix, xl, xc, cd, cm)
	_regex_verifySubTrm_IV_pass("iv");
	_regex_verifySubTrm_IV_fail("ivi");
	_regex_verifySubTrm_IV_fail("ivv");
	_regex_verifySubTrm_IV_fail("viv");
	
	_regex_verifySubTrm_IX_pass("ix");
	_regex_verifySubTrm_IX_pass("xix");
	_regex_verifySubTrm_IX_fail("ixi");
	_regex_verifySubTrm_IX_fail("ixiv");
	_regex_verifySubTrm_IX_fail("ixv");
	
	_regex_verifySubTrm_XL_pass("xl");
	_regex_verifySubTrm_XL_pass("xli");
	_regex_verifySubTrm_XL_pass("xlix");
	_regex_verifySubTrm_XL_pass("cxl");
	_regex_verifySubTrm_XL_fail("xlx");
	_regex_verifySubTrm_XL_fail("xll");
	
	_regex_verifySubTrm_XC_pass("xc");
	_regex_verifySubTrm_XC_pass("cxc");
	_regex_verifySubTrm_XC_pass("xcix");
	_regex_verifySubTrm_XC_pass("xcvi");
	_regex_verifySubTrm_XC_pass("xciv");
	_regex_verifySubTrm_XC_fail("xcx");
	_regex_verifySubTrm_XC_fail("xcxl");
	_regex_verifySubTrm_XC_fail("xcl");
	
	_regex_verifySubTrm_CD_pass("cd");
	_regex_verifySubTrm_CD_pass("cdxl");
	_regex_verifySubTrm_CD_pass("cdxc");
	_regex_verifySubTrm_CD_pass("cdi");
	_regex_verifySubTrm_CD_pass("cdv");
	_regex_verifySubTrm_CD_pass("cdx");
	_regex_verifySubTrm_CD_fail("cdc");
	_regex_verifySubTrm_CD_fail("dcd");
	
	_regex_verifySubTrm_CM_pass("cm");
	_regex_verifySubTrm_CM_pass("cmi");
	_regex_verifySubTrm_CM_pass("cmxc");
	_regex_verifySubTrm_CM_pass("cmxl");
	_regex_verifySubTrm_CM_pass("mcm");
	_regex_verifySubTrm_CM_pass("cmlxvi");
	_regex_verifySubTrm_CM_fail("cmc");
	_regex_verifySubTrm_CM_fail("cmd");
	_regex_verifySubTrm_CM_fail("cmcd");
	
	// write generated roman numerals to a file and compare to a verified file to ensure roman numerals are generated correctly
	file1 = fopen(testFile, "a");
							 
	if(file1 == NULL)
	{
		printf("\r\nError: Failed to create %s", testFile);
	}
	
	for(i = 1; i <= 1000; i++)
	{
		char bfr[16];
		char bfr2[16];
		
		RMN_bfr_init(bfr);
		RMN_bfr_init(bfr2);
			
		RMN_encode_int(i, bfr);
		
		_regex_verifyStr_pass(i, bfr);
		
		sprintf(bfr2, "%d=%s\n", i, bfr);
		fputs(bfr2, file1);
	}

	fclose(file1);
	
	_regex_verifyFile_pass();
						 
	if(remove(testFile) == 0)
	{
		printf("\r\n%s deleted\r\n", testFile);
	}
}


inc_bld_str.sh
#!/bin/bash

# --------------- inc_bld_str.sh log ---------------
#
# 02/06/2017	HSamra	Initial Commit
#
# --------------------------------------------------

#variables:
FILE="bld_num.txt"
NUM=9999 # default value (should never reach this)
STR=""

#code:
#-look for file.  If not found then create it
if [ ! -e $FILE ]
then
	echo creating file...
	> $FILE
fi

#-get number from file
#--without the use of "10#" the number is read as
#--an octal number but will crash at "0008" since
#--it is too large to be an octal number.  Prefix
#--"10#" converts the number into decimal form so
#--we can continue to work with it
while IFS='' read -r line || [[ -n "$line" ]]
do
	NUM="10#$line"
done < $FILE

#-increment
(( NUM++ ))

#--the largest number we want to keep is 9999
if [ $NUM -gt 9999 ] 
then
	(( NUM = 0 ))
fi

#-write new number to $FILE
printf -v STR "%04d" $NUM
echo $STR > $FILE

#-print new number
echo build number: $STR


rmn_calc.c
// --------------- rmn_calc.c Log --------------------
//
// 02/06/2017	HSamra	Initial Commit
// 02/06/2017	HSamra	Adding functionality for string verification
//
// ---------------------------------------------------

#include <regex.h>
#include "common.h"

char rgx_str[] = {"^((m*)((cm?)|((cd?)|(d?)(c{0,3})))((xc?)|((xl?)|(l?)(x{0,3})))((ix?)|((iv?)|(v?)(i{0,3}))))$"};

regex_t regex;
byte rgxSts = 0;

void RMN_bfr_init(char *str)
{
	memset(str, 0x00, sizeof(str));
}

int RMN_init(void)
{
	// compile regex
	if(regcomp(&regex, rgx_str, REG_EXTENDED | REG_ICASE) != 0) // using extended regular expressions and ignoring case
	{
		printf("Could not compile regex\r\n");
		return 0;
	}
	return 1;
}

int RMN_verify_string(char *str)
{
	byte rgxSts = 0;
	
	rgxSts = regexec(&regex, str, 0, NULL, 0);
	
	if(!rgxSts)
	{
		// if match
		return 1;
	}
	else if(rgxSts == REG_NOMATCH)
	{
		// if no match
		return 0;
	}
	else
	{
		// other errors
		char rgxErr[16];
		memset(rgxErr, 0x00, sizeof(rgxErr));
		regerror(rgxSts, &regex, rgxErr, sizeof(rgxErr));
		printf("Regex error: %s\r\n", rgxErr);
	}
	return 0;	
}

void RMN_encode_int(int n, char *str)
{
	while(n > 0)
	{
		if(n >= 1000)
		{
			strcat(str, "M");
			n -= 1000;
		}
		else if(n < 1000 && n >= 900)
		{
			strcat(str, "CM");
			n -= 900;
		}
		else if(n < 900 && n >= 500)
		{
			strcat(str, "D");
			n -= 500;
		}
		else if(n < 500 && n >= 400)
		{
			strcat(str, "CD");
			n -= 400;
		}			
		else if(n < 400 && n >= 100)
		{
			strcat(str, "C");
			n -= 100;
		}
		else if(n < 100 && n >= 90)
		{
			strcat(str, "XC");
			n -= 90;
		}
		else if(n < 90 && n >= 50)
		{
			strcat(str, "L");
			n -= 50;
		}
		else if(n < 50 && n >= 40)
		{
			strcat(str, "XL");
			n -= 40;
		}
		else if(n >= 10)
		{
			strcat(str, "X");
			n -= 10;
		}
		else if(n == 9)
		{
			strcat(str, "IX");
			n -= 9;
		}
		else if(n < 9 && n >= 5)
		{
			strcat(str, "V");
			n -= 5;
		}
		else if(n == 4)
		{
			strcat(str, "IV");
			n -= 4;
		}
		else if(n <= 3)
		{
			strcat(str, "I");
			n -= 1;
		}
		else
		{
			printf("\r\n<ERR: RMN_encode_int>\r\n");
			return;
		}
	}
}

int _decode_char(char chr)
{
	int num = 0;

	switch(toupper(chr))
	{
		case 'I': num = 1; break;
	}
	return num;
}

void RMN_decode_str(int *n, char *str)
{
	*n = _decode_char(str[0]);
}